### ISSUE TITLE
502 Bad gateway of API calls after post form data 

### ISSUE THREAD

**Issue by matthieugouel** [traefik/traefik#8137]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I have 502 Bad request returning from Traefik when I do requests close in time in my API after have uploaded a file via a HTTP POST form data. (https://fastapi.tiangolo.com/tutorial/request-files/) If I wait a few seconds, I don't have the problem. 

### What did you expect to see?

I expect not to have errors when I do a call after uploading a file to my API. I tested the same experiment directly calling the API without Traefik and I don't have any issue. 

### What did you see instead?

I don't have any log from my API but the response of the API call is 502 Bad Request. 

### Output of `traefik version`: (_What version of Traefik are you using?_)

 

```
2.4.8 (but I had the same issue in earlier versions)
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

I made the configurations only with docker compose.
Traefik:

```
    command:
      - "--entrypoints.web.address=:80"
      - "--api=true"
      - "--api.insecure=true"
      - "--providers.docker"
```

On the API:
```
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.api.entrypoints=web"
      - "traefik.http.routers.api.rule=Host(`iris.docker.localhost`) && PathPrefix(`/api`)"
```

### If applicable, please paste the log output in DEBUG level (`--log.level=DEBUG` switch)

```
traefik_1        | time="2021-05-15T17:36:33Z" level=debug msg="'502 Bad Gateway' caused by: EOF"
```

**Comment by matthieugouel** [comment]
I tested this solution (https://github.com/traefik/traefik/issues/3237#issuecomment-514178590) consisting in setting `MaxIdleConnsPerHost = -1` in the configuration of the API in the docker-compose.yml file with no luck : 

```
- "traefik.http.serversTransports.api.maxIdleConnsPerHost=500"
```

**Comment by rtribotte** [comment]
Hello @matthieugouel,

Thanks for your interest in Traefik!

Can you share a complete reproduction case (full configuration) and the full debug log?
For now, it would difficult to assume anything without more information.

Are you sure of your configuration setup, do you effectively reach the expected endpoint on the FastAPI side ([behind proxy doc](https://fastapi.tiangolo.com/advanced/behind-a-proxy/#testing-locally-with-traefik)).

**Comment by matthieugouel** [comment]
Hello @rtribotte, thanks for responding !

So first of all, I have no problem reaching the endpoint behind traefik in normal condition. The only problem I have is when I do API calls directly after uploading a file. If I wait a little (few seconds) after the upload, it working great again.

On the API logs, I never have the problematic call (the one I have a 502 bad gateway response) as if traefik thought that the API was down and took the decision to not transmitting the request to the API.

Also, I don't have this issue if I do calls directly on the API without passing through traefik. 

Finally, I have this problem on testing and production environment. 

Here is an extract of the `docker-compose.yml` file. I don't have any `traefik.yml` file. 

```
  traefik:
    image: "traefik:latest"
    command:
      - "--entrypoints.web.address=:80"
      - "--entrypoints.redis.address=:6379"
      - "--api=true"
      - "--api.insecure=true"
      - "--providers.docker"
    ports:
      - "80:80"
      - "8080:8080"
    volumes:
      - "/var/run/docker.sock:/var/run/docker.sock:ro"
    networks:
      default:
        aliases:
          - "api.docker.localhost"
          - "minio.docker.localhost"

  api:
    build:
      context: .
      dockerfile: "./dockerfiles/Dockerfile-api"
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.api.entrypoints=web"
      - "traefik.http.routers.api.rule=Host(`iris.docker.localhost`) && PathPrefix(`/api`)"

  minio:
    image: "minio/minio:latest"
    command: "server /data"
    labels:
      - "traefik.enable=true"
      - "traefik.http.routers.minio.entrypoints=web"
      - "traefik.http.routers.minio.rule=Host(`minio.docker.localhost`)"
    volumes:
      - "./volumes/minio:/data"
```

The MinIO part is here because the API upload the file that is uploaded in multipart form data to the api. It put it here as well because we see it on the logs 

On the log part, I don't have any logs without the `--log.level=DEBUG`. 
With this option, I don't see any problem beside the : 

```
traefik_1        | time="2021-05-15T17:36:33Z" level=debug msg="'502 Bad Gateway' caused by: EOF"
```

Here is a more detailed extract : 

```
traefik_1        | time="2021-05-17T12:07:18Z" level=debug msg="vulcand/oxy/roundrobin/rr: Forwarding this request to URL" Request="{\"Method\":\"PUT\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/targets-admin/website.csv\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept-Encoding\":[\"identity\"],\"Authorization\":[\"AWS4-HMAC-SHA256 Credential=minioadmin/20210517/local/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=237642957224c90d063a3b3ef50ac23219b8eee8d69e6a07d9cc0d5a87adb25e\"],\"Content-Length\":[\"41\"],\"Content-Md5\":[\"iioSquO2McPDT+/NsmD53w==\"],\"Expect\":[\"100-continue\"],\"User-Agent\":[\"Boto3/1.16.52 Python/3.8.6 Linux/5.10.25-linuxkit Botocore/1.19.52\"],\"X-Amz-Content-Sha256\":[\"4f86ff1afae8eaa34fe6be1a24b3bb75ee8d46dded512a6d832e28553caa49a3\"],\"X-Amz-Date\":[\"20210517T120718Z\"],\"X-Forwarded-Host\":[\"minio.docker.localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"dec6bd32a9be\"],\"X-Real-Ip\":[\"172.20.0.7\"]},\"ContentLength\":41,\"TransferEncoding\":null,\"Host\":\"minio.docker.localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.20.0.7:53566\",\"RequestURI\":\"/targets-admin/website.csv\",\"TLS\":null}" ForwardURL="http://172.20.0.5:9000"
traefik_1        | time="2021-05-17T12:07:18Z" level=debug msg="vulcand/oxy/roundrobin/rr: completed ServeHttp on request" Request="{\"Method\":\"PUT\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/targets-admin/website.csv\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept-Encoding\":[\"identity\"],\"Authorization\":[\"AWS4-HMAC-SHA256 Credential=minioadmin/20210517/local/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=237642957224c90d063a3b3ef50ac23219b8eee8d69e6a07d9cc0d5a87adb25e\"],\"Content-Length\":[\"41\"],\"Content-Md5\":[\"iioSquO2McPDT+/NsmD53w==\"],\"Expect\":[\"100-continue\"],\"User-Agent\":[\"Boto3/1.16.52 Python/3.8.6 Linux/5.10.25-linuxkit Botocore/1.19.52\"],\"X-Amz-Content-Sha256\":[\"4f86ff1afae8eaa34fe6be1a24b3bb75ee8d46dded512a6d832e28553caa49a3\"],\"X-Amz-Date\":[\"20210517T120718Z\"],\"X-Forwarded-Host\":[\"minio.docker.localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"dec6bd32a9be\"],\"X-Real-Ip\":[\"172.20.0.7\"]},\"ContentLength\":41,\"TransferEncoding\":null,\"Host\":\"minio.docker.localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.20.0.7:53566\",\"RequestURI\":\"/targets-admin/website.csv\",\"TLS\":null}"
traefik_1        | time="2021-05-17T12:07:22Z" level=debug msg="'502 Bad Gateway' caused by: EOF"
traefik_1        | time="2021-05-17T12:07:22Z" level=debug msg="vulcand/oxy/roundrobin/rr: completed ServeHttp on request" Request="{\"Method\":\"POST\",\"URL\":{\"Scheme\":\"\",\"Opaque\":\"\",\"User\":null,\"Host\":\"\",\"Path\":\"/api/measurements/\",\"RawPath\":\"\",\"ForceQuery\":false,\"RawQuery\":\"\",\"Fragment\":\"\",\"RawFragment\":\"\"},\"Proto\":\"HTTP/1.1\",\"ProtoMajor\":1,\"ProtoMinor\":1,\"Header\":{\"Accept\":[\"application/json, text/plain, */*\"],\"Accept-Encoding\":[\"gzip, deflate, br\"],\"Accept-Language\":[\"fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7\"],\"Authorization\":[\"Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJzdWIiOiJhZG1pbiIsImV4cCI6MTYyMTMzOTUyNH0.cKeRaTpH-I4jQ-owXZaCt8uAopg2Tx7gtEb0cqXrk8w\"],\"Connection\":[\"keep-alive\"],\"Content-Length\":[\"130\"],\"Content-Type\":[\"application/json;charset=UTF-8\"],\"Origin\":[\"http://localhost:8081\"],\"Referer\":[\"http://localhost:8081/\"],\"Sec-Fetch-Dest\":[\"empty\"],\"Sec-Fetch-Mode\":[\"cors\"],\"Sec-Fetch-Site\":[\"cross-site\"],\"Sec-Gpc\":[\"1\"],\"User-Agent\":[\"Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/90.0.4430.212 Safari/537.36\"],\"X-Forwarded-Host\":[\"iris.docker.localhost\"],\"X-Forwarded-Port\":[\"80\"],\"X-Forwarded-Proto\":[\"http\"],\"X-Forwarded-Server\":[\"dec6bd32a9be\"],\"X-Real-Ip\":[\"172.20.0.1\"]},\"ContentLength\":130,\"TransferEncoding\":null,\"Host\":\"api.docker.localhost\",\"Form\":null,\"PostForm\":null,\"MultipartForm\":null,\"Trailer\":null,\"RemoteAddr\":\"172.20.0.1:61440\",\"RequestURI\":\"/api/measurements/\",\"TLS\":null}"
```
We can see the actual request after the `'502 Bad Gateway' caused by: EOF`. 
I try to give all the necessary information without drowning in the details.

Best, 
Matthieu.

**Comment by tomMoulard** [comment]
Hi @matthieugouel,

Since I don't have your Dockerfile, I could not reproduce your issue. But it seems that it might be an issue of FastAPI [itself](https://github.com/tiangolo/fastapi/issues/2099).

Traefik Already have an issue of that kind `'502 Bad Gateway' caused by: EOF` when, in HTTP2, the backend does not support `keepalive` (#4481).

In this case, you have Two options:
 - Disable traefik's [HTTP2](https://doc.traefik.io/traefik/master/routing/services/#disablehttp2) to remove `keepalive` from traefik (available in the next release of traefik: v2.5).
 - Enable FastAPI support of `keepalive` (using gunicorn for [example](https://github.com/tiangolo/uvicorn-gunicorn-fastapi-docker)).

**Comment by matthieugouel** [comment]
Yes sorry the code of the API is not open-source right know unfortunately. 
I forgot to mention that I'm already using gunicorn with FastAPI using `tiangolo/uvicorn-gunicorn:python3.8` docker image with every parameter to default beside the `MODULE_NAME`. 
```
{"loglevel": "info", "workers": 4, "bind": "0.0.0.0:80", "graceful_timeout": 120, "timeout": 120, "keepalive": 2, "errorlog": "-", "accesslog": "-", "workers_per_core": 1.0, "use_max_workers": null, "host": "0.0.0.0", "port": "80"}
```

I tried to change the value of keepalive and I didn't solve the issue. But I noticed that the time of the 502 bad gateway response is proportional to the value of the keepalive : if I put keepalive to 20s, it takes 20s to respond 502 gateway. 

Note that I've only have this problem after uploading a file through the API. Is it because during the upload the API is not responding and so I have to wait N keepalive seconds for traefik to forward the calls again to the backend server ?

**Comment by matthieugouel** [comment]
Ok I found the issue and it's a little bit hard to explain without all the context. 

Basically, the API upload route was using FastAPI background tasks (https://fastapi.tiangolo.com/tutorial/background-tasks/) in order to transfert the file to S3. But the function of transfering was blocking and blocked the calls that was made during the uploading. 

In order to simplify things I just removed the background task. 

So ... at the end of the day I'm not sure anymore that it was related to Traefik. Sorry for wasting your time and thank you for your help. 

Closing.
